fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/MembershipServiceImpl.java

```diff
@@ -0,0 +1,398 @@
+                final var memberNode = NodeFactory.createURI(member.getFedoraId().getFullId());
+                log.debug("Populating membership history for DirectContainer {}member {}",
+                        memberDeleted ? "deleted " : "", member.getFedoraId());
+                final Instant memberCreated;
+                // Get the creation time from the deleted object if the member is a tombstone
+                if (memberDeleted) {
+                    memberCreated = ((Tombstone) member).getDeletedObject().getCreatedDate();
+                } else {
+                    memberCreated = member.getCreatedDate();
+                }
+                final var memberModified = member.getLastModifiedDate();
+                final var memberEnd = memberDeleted ? memberModified : NO_END_INSTANT;
+
+                // Reduce timeline to just states in effect after the member was created
+                var timelineStream = propertyTimeline.stream()
+                        .filter(e -> e.endDatetime.compareTo(memberCreated) > 0);
+                // If the member was deleted, then reduce timeline to states before the deletion
+                if (memberDeleted) {
+                    timelineStream = timelineStream.filter(e -> e.mementoDatetime.compareTo(memberModified) < 0);
+                }
+                // Index each addition or change to the membership generated by this member
+                timelineStream.forEach(e -> {
+                    // Start time of the membership is the later of member creation or membership resc memento time
+                    indexManager.addMembership(txId, containerId,
+                            generateMembershipTriple(e.membershipResource,
+                                    memberNode, e.hasMemberRelation, e.isMemberOfRelation),
+                            instantMax(memberCreated, e.mementoDatetime),
+                            instantMin(memberEnd, e.endDatetime));
+                });
+            });
+        }
+    }
+
+    private Instant instantMax(final Instant first, final Instant second) {
+        if (first.isAfter(second)) {
+            return first;
+        } else {
+            return second;
+        }
+    }
+
+    private Instant instantMin(final Instant first, final Instant second) {
+        if (first.isBefore(second)) {
+            return first;
+        } else {
+            return second;
+        }
+    }
+
+    /**
+     * Creates a timeline of states for a DirectContainer, tracking changes to its
+     * properties that impact membership.
+     * @param fedoraResc resource subject of the timeline
+     * @return timeline
+     */
+    private List<DirectContainerProperties> makePropertyTimeline(final FedoraResource fedoraResc) {
+        final var entryList = fedoraResc.getTimeMap().getChildren()
+                .sorted((m1, m2) -> m1.getFedoraId().getFullId().compareTo(m1.getFedoraId().getFullId()))
```

Comment:
`m2` is not used.

I think this could also be rewritten like: `.sorted(Comparator.comparing(m -> m.getFedoraId().getFullId()))`

Reply:
Are you just trying to sort the mementos by date? If so, I think they should already be sorted.

Reply:
Yes, I'm trying to sort them by date, I'm happy to remove the sort (and its typo) if we can rely on the versions to be sorted in ascending order. If that's a guarantee, perhaps we should update the documentation for `org.fcrepo.persistence.api.PersistentStorageSession.listVersions(FedoraId)` to state that? Any other theoretic persistence implementations would need to ensure order too.